providers/azure.go

```diff
@@ -72,59 +81,243 @@ func getAzureHeader(accessToken string) http.Header {
-	return email, err
+// Get list of groups user belong to. Filter the desired names of groups (in case of huge group set)
+func (p *AzureProvider) GetGroups(s *sessions.SessionState, f string) (map[string]string, error) {
+	// Azure App Registration requires setting groupMembershipClaims to include group membership in identity token
+	// This option is available through ARM template only.
+	// For details refer to: https://docs.microsoft.com/pl-pl/azure/active-directory/develop/reference-app-manifest
+	if s.IDToken == "" {
+		return map[string]string{}, errors.New("missing id token")
+	}
+
+	type GroupClaims struct {
+		Groups []string `json:"groups"`
+		jwt.StandardClaims
+	}
+
+	claims := &GroupClaims{}
+	jwt.ParseWithClaims(s.IDToken, claims, func(token *jwt.Token) (interface{}, error) {
+		return []byte("empty"), nil
+	})
+
+	groupsMap := make(map[string]string)
+	for _, s := range claims.Groups {
+		groupsMap[s] = s
+	}
+	return groupsMap, nil
+}
+
+// ValidateExemptions checks if we can allow user login dispite group membership returned failure
+func (p *AzureProvider) ValidateExemptions(s *sessions.SessionState) (bool, string) {
+	logger.Printf("ValidateExemptions: validating for %v : %v", s.Email, s.ID)
+	for eAccount, eGroup := range p.ExemptedUsers {
+		if eAccount == s.Email || eAccount == s.Email+":"+s.ID {
+			logger.Printf("ValidateExemptions: \t found '%v' user in exemption list. Returning '%v' group membership", eAccount, eGroup)
+			return true, eGroup
+		}
+	}
+	return false, ""
+}
+
+func (p *AzureProvider) GetLoginURL(redirectURI, state string) string {
+	var a url.URL
+	a = *p.LoginURL
+	params, _ := url.ParseQuery(a.RawQuery)
+	params.Set("client_id", p.ClientID)
+	params.Set("response_type", "id_token code")
+	params.Set("redirect_uri", redirectURI)
+	params.Set("response_mode", "form_post")
+	params.Add("scope", p.Scope)
+	params.Add("state", state)
+	params.Set("prompt", p.ApprovalPrompt)
+	params.Set("nonce", "FIXME")
+	if p.ProtectedResource != nil && p.ProtectedResource.String() != "" {
+		params.Add("resource", p.ProtectedResource.String())
+	}
+	a.RawQuery = params.Encode()
+
+	return a.String()
+}
+
+func (p *AzureProvider) SetGroupRestriction(groups []string) {
+	// Get list of groups (optionally with Group IDs) that ONLY allowed for user
+	// That means even if user has wider group membership, only membership in those groups will be forwarded
+
+	p.PermittedGroups = make(map[string]string)
+	if len(groups) == 0 {
+		return
+	}
+	logger.Printf("Set group restrictions. Allowed groups are:")
+	logger.Printf("\t                *GROUP NAME*      : *GROUP ID*")
+	for _, pGroup := range groups {
+		splittedGroup := strings.Split(pGroup, ":")
+		var groupName string
+		var groupID string
+
+		if len(splittedGroup) == 1 {
+			groupName, groupID = splittedGroup[0], ""
+			p.PermittedGroups[splittedGroup[0]] = ""
+		} else if len(splittedGroup) > 2 {
+			logger.Fatalf("failed to parse '%v'. Too many ':' separators", pGroup)
```

Comment:
`logger.Fatalf` will panic the program IIRC, we shouldn't be panicking if we can't authenticate a user, we should return an internal server error perhaps but not panic the whole program